src/project/project_controller.ts

```diff
@@ -0,0 +1,213 @@
+import { Router, Request, Response, NextFunction,} from 'express';
+import Controller from '../interface/controller_interface';
+import projectModel from './project_model';
+import { Project } from './project_interface';
+import { createProjectSchema, updateProjectStatusSchema, addTeamLeadProjectSchema, addOrRemoveTeamMemberProjectSchema ,validateBody } from '../middleware/validateBody';
+import { authorizeAdmin, authorizeProjectManager, authorizeTeamLead, authorizeTeamMember, authorizeUser } from '../middleware/authorization';
+import { RequestWithUser } from 'interface/requestWithUser';
+
+class ProjectController implements Controller{
+    
+    path: string = '/project';
+    router: Router = Router();
+    private project = projectModel;
+
+    constructor() {
+        this.initializeRoutes()
+    }
+
+    private initializeRoutes(){
+          this.router.post(`${this.path}`, authorizeUser, authorizeAdmin, validateBody(createProjectSchema) ,this.createProject)
+          this.router.delete(`${this.path}/delete/:projectId`, authorizeUser, authorizeAdmin ,this.deleteProject)
+          this.router.get(`${this.path}/get/admin`, authorizeUser, authorizeAdmin, this.getAllProjects)
+          this.router.get(`${this.path}/get/projectManager`, authorizeUser, authorizeProjectManager, this.getProjectsProjectManager)
+          this.router.get(`${this.path}/get/teamLead`, authorizeUser, authorizeTeamLead, this.getProjectsTeamLead)
+          this.router.get(`${this.path}/get/teamMember`, authorizeUser, authorizeTeamMember, this.getProjectsTeamMembers)
+          this.router.put(`${this.path}/add/teamLead/:projectId`, authorizeUser, authorizeProjectManager, validateBody(addTeamLeadProjectSchema), this.addTeamLead)
+          this.router.put(`${this.path}/teamMember/:projectId`, authorizeUser, authorizeProjectManager, validateBody(addOrRemoveTeamMemberProjectSchema), this.addTeamMember)
+          this.router.put(`${this.path}/teamMember/remove/:projectId`, authorizeUser, authorizeProjectManager, validateBody(addOrRemoveTeamMemberProjectSchema), this.removeTeamMember)
+          this.router.put(`${this.path}/status/:projectId`, authorizeUser, authorizeProjectManager, validateBody(updateProjectStatusSchema) , this.updateProjectStatus)
+          this.router.get(`${this.path}/members/:projectId`, authorizeUser, authorizeTeamLead , this.getAllMembers)
```

Comment:
Overall it looks good. But personally, I would change it to a couple of method calls. 
`configureAdminRoutes()`
`configureProjectManagerRoutes()`
`configureTeamLeadRoutes()`
`configureTeamMemberRoutes()`

Reply:
These methods you mentioned will be middlewares?

Reply:
No, these methods will be the methods of the controller class we are implementing and will divide the routes based on user roles for better identification and review in the future. Right now, everything is in one place so it is causing a readability issue for what is related to the team lead and what's related to the project manager.